File: crates/alpaca-option/src/pricing.rs
use crate::error::{OptionError, OptionResult};
use crate::numeric::{brent_solve, normal_cdf, normal_pdf};
use crate::types::{
    BlackScholesImpliedVolatilityInput, BlackScholesInput, Greeks, OptionContract, OptionRight,
};

fn ensure_finite(name: &str, value: f64) -> OptionResult<()> {
    if value.is_finite() {
        Ok(())
    } else {
        Err(OptionError::new(
            "invalid_pricing_input",
            format!("{name} must be finite: {value}"),
        ))
    }
}

fn ensure_positive(name: &str, value: f64) -> OptionResult<()> {
    ensure_finite(name, value)?;
    if value > 0.0 {
        Ok(())
    } else {
        Err(OptionError::new(
            "invalid_pricing_input",
            format!("{name} must be greater than zero: {value}"),
        ))
    }
}

fn parse_option_right(option_right: &str) -> OptionResult<OptionRight> {
    OptionRight::from_str(option_right).map_err(|_| {
        OptionError::new(
            "invalid_pricing_input",
            format!("invalid option right: {option_right}"),
        )
    })
}

fn validate_black_scholes_inputs(input: &BlackScholesInput) -> OptionResult<()> {
    ensure_positive("spot", input.spot)?;
    ensure_positive("strike", input.strike)?;
    ensure_positive("years", input.years)?;
    ensure_finite("rate", input.rate)?;
    ensure_finite("dividend_yield", input.dividend_yield)?;
    ensure_positive("volatility", input.volatility)?;
    Ok(())
}

fn d1_d2(input: &BlackScholesInput) -> (f64, f64) {
    let sqrt_years = input.years.sqrt();
    let sigma_sqrt_t = input.volatility * sqrt_years;
    let d1 = ((input.spot / input.strike).ln()
        + (input.rate - input.dividend_yield + 0.5 * input.volatility * input.volatility)
            * input.years)
        / sigma_sqrt_t;
    let d2 = d1 - sigma_sqrt_t;
    (d1, d2)
}

fn price_black_scholes_core(input: &BlackScholesInput) -> f64 {
    let (d1, d2) = d1_d2(input);
    let discount_spot = (-input.dividend_yield * input.years).exp();
    let discount_strike = (-input.rate * input.years).exp();

    match input.option_right {
        OptionRight::Call => {
            input.spot * discount_spot * normal_cdf(d1)
                - input.strike * discount_strike * normal_cdf(d2)
        }
        OptionRight::Put => {
            input.strike * discount_strike * normal_cdf(-d2)
                - input.spot * discount_spot * normal_cdf(-d1)
        }
    }
}

fn discounted_forward_minus_strike(input: &BlackScholesInput) -> f64 {
    input.spot * (-input.dividend_yield * input.years).exp()
        - input.strike * (-input.rate * input.years).exp()
}

fn european_no_arbitrage_lower_bound(input: &BlackScholesInput) -> f64 {
    let parity = discounted_forward_minus_strike(input);
    match input.option_right {
        OptionRight::Call => parity.max(0.0),
        OptionRight::Put => (-parity).max(0.0),
    }
}

pub fn price_black_scholes(input: &BlackScholesInput) -> OptionResult<f64> {
    validate_black_scholes_inputs(input)?;
    Ok(price_black_scholes_core(input))
}

pub fn greeks_black_scholes(input: &BlackScholesInput) -> OptionResult<Greeks> {
    validate_black_scholes_inputs(input)?;
    let (d1, d2) = d1_d2(input);
    let sqrt_years = input.years.sqrt();
    let sigma_sqrt_t = input.volatility * sqrt_years;
    let exp_minus_qt = (-input.dividend_yield * input.years).exp();
    let exp_minus_rt = (-input.rate * input.years).exp();
    let nd1 = normal_cdf(d1);
    let nd2 = normal_cdf(d2);
    let n_minus_d1 = normal_cdf(-d1);
    let n_minus_d2 = normal_cdf(-d2);
    let phi_d1 = normal_pdf(d1);

    let delta = match input.option_right {
        OptionRight::Call => exp_minus_qt * nd1,
        OptionRight::Put => -exp_minus_qt * n_minus_d1,
    };
    let gamma = exp_minus_qt * phi_d1 / (input.spot * sigma_sqrt_t);
    let vega = input.spot * exp_minus_qt * phi_d1 * sqrt_years / 100.0;
    let theta_annual = match input.option_right {
        OptionRight::Call => {
            -input.spot * exp_minus_qt * phi_d1 * input.volatility / (2.0 * sqrt_years)
                - input.rate * input.strike * exp_minus_rt * nd2
                + input.dividend_yield * input.spot * exp_minus_qt * nd1
        }
        OptionRight::Put => {
            -input.spot * exp_minus_qt * phi_d1 * input.volatility / (2.0 * sqrt_years)
                + input.rate * input.strike * exp_minus_rt * n_minus_d2
                - input.dividend_yield * input.spot * exp_minus_qt * n_minus_d1
        }
    };
    let theta = theta_annual / 365.0;
    let rho = match input.option_right {
        OptionRight::Call => input.strike * input.years * exp_minus_rt * nd2,
        OptionRight::Put => -input.strike * input.years * exp_minus_rt * n_minus_d2,
    };

    Ok(Greeks {
        delta,
        gamma,
        vega,
        theta,
        rho,
    })
}

pub fn intrinsic_value(spot: f64, strike: f64, option_right: &str) -> OptionResult<f64> {
    ensure_finite("spot", spot)?;
    ensure_positive("strike", strike)?;
    let option_right = parse_option_right(option_right)?;

    Ok(match option_right {
        OptionRight::Call => (spot - strike).max(0.0),
        OptionRight::Put => (strike - spot).max(0.0),
    })
}

pub fn extrinsic_value(option_price: f64, spot: f64, strike: f64, option_right: &str) -> OptionResult<f64> {
    ensure_finite("option_price", option_price)?;
    Ok((option_price - intrinsic_value(spot, strike, option_right)?).max(0.0))
}

pub fn contract_extrinsic_value(option_price: f64, spot: f64, contract: &OptionContract) -> OptionResult<f64> {
    extrinsic_value(option_price, spot, contract.strike, contract.option_right.as_str())
}

pub fn implied_volatility_from_price(
    input: &BlackScholesImpliedVolatilityInput,
) -> OptionResult<f64> {
    ensure_finite("target_price", input.target_price)?;
    let black_scholes_input = BlackScholesInput {
        spot: input.spot,
        strike: input.strike,
        years: input.years,
        rate: input.rate,
        dividend_yield: input.dividend_yield,
        volatility: 0.2,
        option_right: input.option_right.clone(),
    };
    validate_black_scholes_inputs(&black_scholes_input)?;
    let minimum_price = european_no_arbitrage_lower_bound(&black_scholes_input);
    if input.target_price + 1e-12 < minimum_price {
        return Err(OptionError::new(
            "invalid_pricing_input",
            format!(
                "target_price is below discounted no-arbitrage lower bound: {} < {minimum_price}",
                input.target_price
            ),
        ));
    }

    let lower_bound = input.lower_bound.unwrap_or(0.0001);
    let upper_bound = input.upper_bound.unwrap_or(5.0);
    ensure_positive("lower_bound", lower_bound)?;
    ensure_positive("upper_bound", upper_bound)?;
    if lower_bound >= upper_bound {
        return Err(OptionError::new(
            "invalid_pricing_input",
            format!("lower_bound must be less than upper_bound: {lower_bound} >= {upper_bound}"),
        ));
    }

    brent_solve(
        lower_bound,
        upper_bound,
        |volatility| price_black_scholes_core(&BlackScholesInput {
            volatility,
            ..black_scholes_input.clone()
        }) - input.target_price,
        input.tolerance,
        input.max_iterations,
    )
}

File: packages/alpaca-option/src/pricing.ts
import * as optionContract from './contract';
import { fail } from './error';
import { brentSolve, normalCdf, normalPdf } from './numeric';
import type {
  BlackScholesImpliedVolatilityInput,
  BlackScholesInput,
  Greeks,
  OptionRight,
} from './types';

type ContractInput = optionContract.ContractInput;

function ensureFinite(name: string, value: number): void {
  if (!Number.isFinite(value)) {
    fail('invalid_pricing_input', `${name} must be finite: ${value}`);
  }
}

function ensurePositive(name: string, value: number): void {
  ensureFinite(name, value);
  if (value <= 0) {
    fail('invalid_pricing_input', `${name} must be greater than zero: ${value}`);
  }
}

function validateOptionRight(optionRight: OptionRight): OptionRight {
  if (optionRight !== 'call' && optionRight !== 'put') {
    fail('invalid_pricing_input', `invalid option right: ${optionRight}`);
  }
  return optionRight;
}

function validateBlackScholesInput(input: BlackScholesInput): OptionRight {
  ensurePositive('spot', input.spot);
  ensurePositive('strike', input.strike);
  ensurePositive('years', input.years);
  ensureFinite('rate', input.rate);
  ensureFinite('dividendYield', input.dividendYield);
  ensurePositive('volatility', input.volatility);
  return validateOptionRight(input.optionRight);
}

function d1d2(input: BlackScholesInput): { d1: number; d2: number } {
  const sqrtYears = Math.sqrt(input.years);
  const sigmaSqrtT = input.volatility * sqrtYears;
  const d1 = (Math.log(input.spot / input.strike)
    + (input.rate - input.dividendYield + 0.5 * input.volatility * input.volatility) * input.years)
    / sigmaSqrtT;
  return { d1, d2: d1 - sigmaSqrtT };
}

function priceBlackScholesCore(input: BlackScholesInput): number {
  const { d1, d2 } = d1d2(input);
  const discountSpot = Math.exp(-input.dividendYield * input.years);
  const discountStrike = Math.exp(-input.rate * input.years);

  if (input.optionRight === 'call') {
    return input.spot * discountSpot * normalCdf(d1)
      - input.strike * discountStrike * normalCdf(d2);
  }
  return input.strike * discountStrike * normalCdf(-d2)
    - input.spot * discountSpot * normalCdf(-d1);
}

function discountedForwardMinusStrike(input: Pick<BlackScholesInput, 'spot' | 'strike' | 'years' | 'rate' | 'dividendYield'>): number {
  return input.spot * Math.exp(-input.dividendYield * input.years)
    - input.strike * Math.exp(-input.rate * input.years);
}

function europeanNoArbitrageLowerBound(input: Pick<BlackScholesInput, 'spot' | 'strike' | 'years' | 'rate' | 'dividendYield' | 'optionRight'>): number {
  const parity = discountedForwardMinusStrike(input);
  return input.optionRight === 'call' ? Math.max(parity, 0) : Math.max(-parity, 0);
}

export function priceBlackScholes(input: BlackScholesInput): number {
  const optionRight = validateBlackScholesInput(input);
  return priceBlackScholesCore({ ...input, optionRight });
}

export function greeksBlackScholes(input: BlackScholesInput): Greeks {
  const optionRight = validateBlackScholesInput(input);
  const { d1, d2 } = d1d2(input);
  const sqrtYears = Math.sqrt(input.years);
  const sigmaSqrtT = input.volatility * sqrtYears;
  const expMinusQt = Math.exp(-input.dividendYield * input.years);
  const expMinusRt = Math.exp(-input.rate * input.years);
  const nd1 = normalCdf(d1);
  const nd2 = normalCdf(d2);
  const nMinusD1 = normalCdf(-d1);
  const nMinusD2 = normalCdf(-d2);
  const phiD1 = normalPdf(d1);

  const delta = optionRight === 'call' ? expMinusQt * nd1 : -expMinusQt * nMinusD1;
  const gamma = expMinusQt * phiD1 / (input.spot * sigmaSqrtT);
  const vega = input.spot * expMinusQt * phiD1 * sqrtYears / 100;
  const thetaAnnual = optionRight === 'call'
    ? -input.spot * expMinusQt * phiD1 * input.volatility / (2 * sqrtYears)
      - input.rate * input.strike * expMinusRt * nd2
      + input.dividendYield * input.spot * expMinusQt * nd1
    : -input.spot * expMinusQt * phiD1 * input.volatility / (2 * sqrtYears)
      + input.rate * input.strike * expMinusRt * nMinusD2
      - input.dividendYield * input.spot * expMinusQt * nMinusD1;
  const theta = thetaAnnual / 365;
  const rho = optionRight === 'call'
    ? input.strike * input.years * expMinusRt * nd2
    : -input.strike * input.years * expMinusRt * nMinusD2;

  return { delta, gamma, vega, theta, rho };
}

export function intrinsicValue(spot: number, strike: number, optionRight: OptionRight): number {
  ensureFinite('spot', spot);
  ensurePositive('strike', strike);
  validateOptionRight(optionRight);
  return optionRight === 'call'
    ? Math.max(spot - strike, 0)
    : Math.max(strike - spot, 0);
}

export function extrinsicValue(
  optionPrice: number,
  spot: number,
  strike: number,
  optionRight: OptionRight,
): number {
  ensureFinite('optionPrice', optionPrice);
  return Math.max(optionPrice - intrinsicValue(spot, strike, optionRight), 0);
}

export function contractExtrinsicValue(
  optionPriceInput: number | string | null,
  spotInput: number | string | null,
  contractInput?: ContractInput,
): number | null {
  const optionPrice = typeof optionPriceInput === 'number' ? optionPriceInput : Number(optionPriceInput);
  const spot = typeof spotInput === 'number' ? spotInput : Number(spotInput);
  const contract = optionContract.canonicalContract(contractInput);
  if (!Number.isFinite(optionPrice) || !Number.isFinite(spot) || !contract) {
    return null;
  }

  return extrinsicValue(optionPrice, spot, contract.strike, contract.option_right);
}

export function impliedVolatilityFromPrice(input: BlackScholesImpliedVolatilityInput): number {
  ensureFinite('targetPrice', input.targetPrice);
  validateBlackScholesInput({ ...input, volatility: 0.2 });

  const minimumPrice = europeanNoArbitrageLowerBound({
    spot: input.spot,
    strike: input.strike,
    years: input.years,
    rate: input.rate,
    dividendYield: input.dividendYield,
    optionRight: input.optionRight,
  });
  if (input.targetPrice + 1e-12 < minimumPrice) {
    fail(
      'invalid_pricing_input',
      `targetPrice is below discounted no-arbitrage lower bound: ${input.targetPrice} < ${minimumPrice}`,
    );
  }

  const lowerBound = input.lowerBound ?? 0.0001;
  const upperBound = input.upperBound ?? 5.0;
  ensurePositive('lowerBound', lowerBound);
  ensurePositive('upperBound', upperBound);
  if (lowerBound >= upperBound) {
    fail('invalid_pricing_input', `lowerBound must be less than upperBound: ${lowerBound} >= ${upperBound}`);
  }

  return brentSolve(
    lowerBound,
    upperBound,
    (volatility) => priceBlackScholesCore({ ...input, volatility }) - input.targetPrice,
    input.tolerance,
    input.maxIterations,
  );
}
